### PR TITLE
feat: add network connection encryption function

### PR DESCRIPTION
**Pull request by mmcshinsky-bitgo**
* Add ability to encrypt connection POST data per partner spec
* Additional automations to add encryption into sdk network methods to come as integrations are verified fully.
* Partner public keys to be added when ready for customer use

PX-3786

### CODE REVIEW

File: modules/sdk-core/src/bitgo/trading/index.ts [hunk 1]
```diff
@@ -1,2 +1,3 @@
+export * from './network';
```
[the remark this answers is not in the document]
Comment (mmcshinsky-bitgo) [reply modules/sdk-core/src/bitgo/trading/index.ts:1]: It has the potential to be moved out. Something to look into as a backlog wishlist.

File: modules/sdk-core/test/unit/bitgo/trading/network/encrypt.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      type: 'spki',
+      format: 'pem',
+    })
+    .toString();
+
+  it('should encrypt', async () => {
+    const password = 'password';
+    const encrypted = await encryptRsaWithAesGcm(publicKeyString, password);
+
+    assert(encrypted);
```
[the remark this answers is not in the document]
Comment (mmcshinsky-bitgo) [reply modules/sdk-core/test/unit/bitgo/trading/network/encrypt.ts]: Tests now added.

File: modules/express/src/clientRoutes.ts [hunk 3]
```diff
@@ -319,6 +322,29 @@ function handleREST(req: express.Request, res: express.Response, next: express.N
   return redirectRequest(bitgo, method, bitgoURL, req, next);
 }
 
+/**
```
Comment (mmcshinsky-bitgo) [inline modules/express/src/clientRoutes.ts:325]: Purposefully replicates req.query portion of handleREST

File: modules/express/src/clientRoutes.ts [hunk 4]
```diff
@@ -1449,6 +1554,21 @@ export function setupAPIRoutes(app: express.Application, config: Config): void {
...
+  );
+
+  app.use(
+    '/api/:namespace/v[12]/enterprises/:enterpriseId/*',
+    parseBody,
+    prepareBitGo(config),
+    promiseWrapper(handleMicroserviceReq)
+  );
+  app.use('/api/:namespace/v[12]/*', parseBody, prepareBitGo(config), promiseWrapper(handleMicroserviceReq));
+
```
Comment (mmcshinsky-bitgo) [inline modules/express/src/clientRoutes.ts:1571]: Adding these `.use`s allows for these calls to be properly handled for namespaced apis, not only the /api/v[12] that were only being handled previously.

Comment (mmcshinsky-bitgo) [reply modules/express/src/clientRoutes.ts:1571]: https://github.com/BitGo/BitGoJS/assets/100626066/2655f27b-9eca-494a-906e-e1022ded4502

Comment (mmcshinsky-bitgo) [reply modules/express/src/clientRoutes.ts:1571]: https://github.com/BitGo/BitGoJS/assets/100626066/b5a72e0a-588d-41e3-a6a9-30bc030f2a4a

File: modules/express/src/clientRoutes.ts [hunk 5]
```diff
@@ -1449,6 +1545,21 @@ export function setupAPIRoutes(app: express.Application, config: Config): void {
...
+    promiseWrapper(handleNetworkV1EnterpriseClientConnections)
+  );
+
+  app.use(
+    '/api/:namespace/v[12]/enterprises/:enterpriseId/*',
+    parseBody,
+    prepareBitGo(config),
+    promiseWrapper(handleMicroserviceReq)
+  );
+  app.use('/api/:namespace/v[12]/*', parseBody, prepareBitGo(config), promiseWrapper(handleMicroserviceReq));
```
Comment (zahin-mohammad) [inline modules/express/src/clientRoutes.ts:1561]: This is a dangerous catch all. A lot of express routes contain private keys and wallet password which we do NOT want coming to wallet platform. This makes it much easier for clients to accidentally leak their data if they for ex mistype a url between express and WP.

Comment (zahin-mohammad) [reply modules/express/src/clientRoutes.ts:1561]: Why do we need to introduce this change in this PR?

**Review by zahin-mohammad — CHANGES_REQUESTED** [review]
the netwrok encrypt/decrypt lgtm, but i have some concerns on the catch all. Let's discuss more in the wallets team chat.

File: modules/sdk-core/src/bitgo/trading/network/types.ts [hunk 6]
```diff
@@ -33,6 +34,18 @@ export type GetNetworkBalancesResponse = {
...
+  ids?: string[];
+  names?: string[];
+  institutionIds?: string[];
+  institutionIdentifiers?: string[];
+  active?: boolean;
+};
+
+export type GetNetworkPartnersResponse = {
+  partners: NetworkPartner[];
+};
```
[the remark this answers is not in the document]
Comment (mmcshinsky-bitgo) [reply modules/sdk-core/src/bitgo/trading/network/types.ts:47]: Thanks. Ticket for public types currently tracked in different Epic.

File: modules/express/src/clientRoutes.ts (hunk 5 above)
Comment (mmcshinsky-bitgo) [reply modules/express/src/clientRoutes.ts]: As well as:
```
  app.use('/api/v2/user/*', parseBody, prepareBitGo(config), promiseWrapper(handleV2UserREST));
  app.use('/api/v2/:coin/*', parseBody, prepareBitGo(config), promiseWrapper(handleV2CoinSpecificREST));
```

Comment (mmcshinsky-bitgo) [reply modules/express/src/clientRoutes.ts]: We could move them into:
```
  // everything else should use the proxy handler
  if (!config.disableProxy) {
  ```

Comment (mmcshinsky-bitgo) [reply modules/express/src/clientRoutes.ts]: Could move into the `disableProxy` function. This would the use of `disableProxy` as well.

Comment (mmcshinsky-bitgo) [reply modules/express/src/clientRoutes.ts]: Update: my code was flawed, so I was able to remove the added middleware and move the namespace catch into disableProxy.

File: modules/express/src/clientRoutes.ts [hunk 7]
```diff
@@ -1002,32 +1070,47 @@ function handleV2CoinSpecificREST(req: express.Request, res: express.Response, n
...
+  }
+
   // something has presumably gone wrong
   next();
 }
 
 async function handleProxyReq(req: express.Request, res: express.Response, next: express.NextFunction) {
-  const fullUrl = req.bitgo.microservicesUrl(req.url);
-  if (req.url && (/^\/api.*$/.test(req.url) || /^\/oauth\/token.*$/.test(req.url))) {
+  const fullUrl = req.bitgo.microservicesUrl(req.originalUrl);
```
Comment (mmcshinsky-bitgo) [inline modules/express/src/clientRoutes.ts:1108]: https://expressjs.com/en/5x/api.html#req.originalUrl

**Review by zahin-mohammad — DISMISSED** [review]
Discussed on slack

File: modules/sdk-core/src/bitgo/trading/network/encrypt.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+ * Provided an X.509/ OpenSSL PEM public key, and a string of text to encrypt,
+ * This function will
+ * 1. Generate a random 256-bit key
+ * 2. Encrypt the text using AES-GCM with the generated key
+ * 3. Encrypt the generated key using RSA-OAEP with the provided public key
+ * 4. Return the encrypted key and the encrypted text in the format `${encryptedKey}\n${encryptedText}`
+ *
+ * @param callWithRetryServiceContext - service dependent context @see {@link CallWithRetryServiceContext}
+ * @param callWithRetryCallContext - call dependent context @see {@link CallWithRetryCallContext<Params, Response>}
+ * @params callWithRetryParams - params to invoke the call @see {@link CallWithRetryParams<Params>}
```
Comment (domoneill18) [inline modules/sdk-core/src/bitgo/trading/network/encrypt.ts:96]:
```suggestion
 * @param publicKey - RSA Public Key
 * @param text - text to encrypt
```

File: modules/sdk-core/src/bitgo/trading/network/encrypt.ts [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+ * @returns TE.TaskEither<Error, Response['body']> from call
```
Comment (domoneill18) [inline modules/sdk-core/src/bitgo/trading/network/encrypt.ts:97]:
```suggestion
 * @returns - encrypted text string
```

File: modules/sdk-core/src/bitgo/trading/network/encrypt.ts [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+  const authTag = cipher.getAuthTag();
+
+  return Buffer.concat([version, salt, iv, encrypted, authTag]).toString('base64');
+}
+
+export async function _decryptAesGcm(secret: string | Buffer, encryptedText: string): Promise<string> {
+  const data = Buffer.from(encryptedText, 'base64');
+
+  // const version = data.slice(0, 1);
```
[the remark this answers is not in the document]
Comment (domoneill18) [reply modules/sdk-core/src/bitgo/trading/network/encrypt.ts:42]:
```suggestion
    const version = message.slice(0, 1);
    if (version.readInt8() !== 1) {
      throw new Error('Unknown encryption version');
    }
```

[the remark this answers is not in the document]
Comment (mmcshinsky-bitgo) [reply modules/sdk-core/src/bitgo/trading/network/encrypt.ts:42]: Forgot to include. Thanks

**Review by domoneill18 — APPROVED** [review]

**Review by zahin-mohammad — APPROVED** [review]

**Pull request merged**